File: armour/src/main/java/com/yangzhiwen/armour/ArmourInstrumentation.java
package com.yangzhiwen.armour;

import android.app.Activity;
import android.app.Instrumentation;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;

import com.yangzhiwen.armour.ext.Hacker;


/**
 * use java，because kotlin Instrumentation execStartActivity can not invoke by system
 * Created by yangzhiwen on 17/8/18.
 */
public class ArmourInstrumentation extends Instrumentation {

    private Instrumentation base;
    private Armour armour;

    public ArmourInstrumentation(Armour armour, Instrumentation base) {
        this.base = base;
        this.armour = armour;
    }

    public ActivityResult execStartActivity(Context who, IBinder contextThread, IBinder token, Activity target,
                                            Intent intent, int requestCode, Bundle options) {
        intent = armour.getArmourHacker().execStartActivity(who, intent);
        return execBaseStartActivity(who, contextThread, token, target, intent, requestCode, options);
    }

    public ActivityResult execBaseStartActivity(
            Context who, IBinder contextThread, IBinder token, Activity target,
            Intent intent, int requestCode, Bundle options) {
        Class[] parameterTypes = {Context.class, IBinder.class, IBinder.class, Activity.class, Intent.class,
                int.class, Bundle.class};
        try {
            return (ActivityResult) Hacker.Companion
                    .on(Instrumentation.class)
                    .method("execStartActivity", parameterTypes)
                    .invoke(base, who, contextThread, token, target, intent, requestCode, options);
        } catch (Exception e) {
            return null;
        }
    }

    @Override
    public Activity newActivity(ClassLoader cl, String className, Intent intent) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
        Activity activity = armour.getArmourHacker().newActivity(className, intent);
        if (activity != null) return activity;
        return super.newActivity(cl, className, intent);
    }

    @Override
    public void callActivityOnCreate(Activity activity, Bundle icicle) {
        armour.getArmourHacker().callActivityOnCreate(activity);
        super.callActivityOnCreate(activity, icicle);
    }
}

File: app/src/main/java/com/yangzhiwen/navigator/ArmourInstrumentation.kt
package com.yangzhiwen.navigator

import android.app.Activity
import android.app.Instrumentation
import android.content.Intent
import android.content.res.Resources
import android.view.ContextThemeWrapper
import com.yangzhiwen.armour.Armour

/**
 * Created by yangzhiwen on 17/8/15.
 */
class ArmourInstrumentation(baseArg: Instrumentation, armourArg: Armour) : Instrumentation() {

    val base = baseArg
    val armour = armourArg


    override fun newActivity(cl: ClassLoader?, className: String?, intent: Intent?): Activity {
        try {
            println("load activity $className")
            if(className == "com.yangzhiwen.navigator.ProxyActivity") throw IllegalArgumentException("ddd")
            cl?.loadClass(className)
        } catch (e: Exception) {
            val curClassLoader = armour.getPlugin("user_center")?.classloader
            if (curClassLoader != null) {
                println("...........dazz")
                val activity = base.newActivity(curClassLoader, "com.yangzhiwen.demo.CenterActivity", intent)
                val oldResources = App.res ?: return super.newActivity(cl, className, intent)
                val newAssetManager = oldResources.assets.javaClass.newInstance()
                val addAssetPathMethod = oldResources.assets.javaClass.getDeclaredMethod("addAssetPath", String::class.java)
                addAssetPathMethod.isAccessible = true
                val temp = addAssetPathMethod.invoke(newAssetManager, App.path) as Int
                if (temp == 0) {
                    println("addAssetPath return 0")
                }

                val newR = Resources(newAssetManager, oldResources.displayMetrics, oldResources.configuration)

                try {
                    // for 4.1+
                    ReflectUtil.setField(ContextThemeWrapper::class.java, activity, "mResources", newR)
                } catch (ignored: Exception) {
                    ignored.printStackTrace()
                    // ignored.
                }
                return activity
            }
        }

        return super.newActivity(cl, className, intent)

    }
}